File: pc3/bd_principal.py
# bd_principal.py - base de datos principal - pc3 (persistencia)
#
# este servicio es la bd principal del sistema. hace dos cosas:
#   1. recibe datos procesados de la analítica (pull) y los guarda en sqlite
#   2. responde consultas del servicio de monitoreo (rep)
#
# si este servicio se cae, la analítica activa el enmascaramiento de fallos
# y el monitoreo se conecta automáticamente a la bd réplica en pc2.
#
# autores: miguel angel acuña, juan david acuña, y samuel felipe manrique - sistemas distribuidos 2026-10

import zmq
import json
import sqlite3
import threading
import time
from datetime import datetime, timezone

# ============================================================
# CONFIGURACIÓN DE RED
# ============================================================
BD_IP = "10.43.99.183"   # PC3 - esta máquina
PUERTO_PULL = 5570        # Puerto para recibir datos de analítica
PUERTO_REP = 5571         # Puerto para consultas del monitoreo

BD_ARCHIVO = "trafico.db"   # Nombre del archivo SQLite

# TODO: Implementar CurveZMQ para la entrega final


def timestamp_ahora():
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


def crear_tablas():
    # creo las tablas de la bd si no existen.
    # son las mismas tablas que en la bd réplica.
    conn = sqlite3.connect(BD_ARCHIVO)
    cursor = conn.cursor()

    # Tabla principal de eventos de tráfico
    cursor.execute("""
        CREATE TABLE IF NOT EXISTS eventos_trafico (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            interseccion TEXT,
            tipo_sensor TEXT,
            datos_sensor TEXT,
            estado_trafico TEXT,
            Q REAL,
            Vp REAL,
            D REAL,
            timestamp_procesado TEXT,
            timestamp_insercion TEXT
        )
    """)

    # Tabla de historial de semáforos
    cursor.execute("""
        CREATE TABLE IF NOT EXISTS estados_semaforo (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            interseccion TEXT,
            estado_anterior TEXT,
            estado_nuevo TEXT,
            duracion_verde INTEGER,
            motivo TEXT,
            timestamp TEXT
        )
    """)

    # Tabla de acciones de control
    cursor.execute("""
        CREATE TABLE IF NOT EXISTS acciones_control (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            interseccion TEXT,
            tipo_accion TEXT,
            detalles TEXT,
            timestamp TEXT
        )
    """)

    conn.commit()
    conn.close()
    print(f"[BD-PRINCIPAL] Base de datos '{BD_ARCHIVO}' lista")


def guardar_evento(registro):
    # guarda un evento procesado en la bd usando sql crudo.
    conn = sqlite3.connect(BD_ARCHIVO)
    cursor = conn.cursor()

    cursor.execute("""
        INSERT INTO eventos_trafico 
        (interseccion, tipo_sensor, datos_sensor, estado_trafico, Q, Vp, D,
         timestamp_procesado, timestamp_insercion)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
    """, (
        registro.get("interseccion", ""),
        registro.get("tipo_sensor", ""),
        json.dumps(registro.get("datos_sensor", {})),
        registro.get("estado_trafico", ""),
        registro.get("Q", 0),
        registro.get("Vp", 0),
        registro.get("D", 0),
        registro.get("timestamp_procesado", ""),
        timestamp_ahora()
    ))

    conn.commit()
    total = cursor.execute("SELECT COUNT(*) FROM eventos_trafico").fetchone()[0]
    conn.close()

    return total


def hilo_recibir_datos(contexto):
    # este hilo recibe datos procesados de la analítica (pull).
    # cada dato que llega lo guardo en la bd sqlite.
    socket = contexto.socket(zmq.PULL)
    socket.bind(f"tcp://{BD_IP}:{PUERTO_PULL}")

    poller = zmq.Poller()
    poller.register(socket, zmq.POLLIN)

    print(f"[BD-PRINCIPAL] PULL esperando datos en tcp://{BD_IP}:{PUERTO_PULL}")

    while True:
        # Espero hasta 2 segundos
        eventos = dict(poller.poll(2000))

        if socket not in eventos:
            continue

        # Recibo el registro como diccionario
        registro = socket.recv_json()

        # Lo guardo en la BD
        total = guardar_evento(registro)

        inter = registro.get("interseccion", "?")
        estado = registro.get("estado_trafico", "?")
        tipo = registro.get("tipo_sensor", "?")
        print(f"[BD-PRINCIPAL] Guardado: {inter} | {tipo} | {estado} | Total: {total}")


def hilo_consultas(contexto):
    # este hilo responde consultas del servicio de monitoreo (rep).
    # hace selects directos a la bd sqlite y devuelve los resultados.
    socket = contexto.socket(zmq.REP)
    socket.bind(f"tcp://{BD_IP}:{PUERTO_REP}")

    poller = zmq.Poller()
    poller.register(socket, zmq.POLLIN)

    print(f"[BD-PRINCIPAL] REP esperando consultas en tcp://{BD_IP}:{PUERTO_REP}")

    while True:
        eventos = dict(poller.poll(2000))

        if socket not in eventos:
            continue

        # Recibo la consulta del monitoreo
        consulta = socket.recv_json()
        tipo = consulta.get("tipo", "")

        print(f"[BD-PRINCIPAL] Consulta: {tipo}")

        conn = sqlite3.connect(BD_ARCHIVO)
        cursor = conn.cursor()

        if tipo == "CONSULTA_HISTORICA":
            # Busco eventos entre dos fechas (útil para horas pico)
            inicio = consulta.get("fecha_inicio", "")
            fin = consulta.get("fecha_fin", "")

            cursor.execute("""
                SELECT interseccion, tipo_sensor, estado_trafico, Q, Vp, D,
                       timestamp_procesado
                FROM eventos_trafico
                WHERE timestamp_procesado BETWEEN ? AND ?
                ORDER BY timestamp_procesado DESC
                LIMIT 100
            """, (inicio, fin))

            resultados = []
            for fila in cursor.fetchall():
                resultados.append({
                    "interseccion": fila[0], "tipo_sensor": fila[1],
                    "estado_trafico": fila[2], "Q": fila[3],
                    "Vp": fila[4], "D": fila[5], "timestamp": fila[6]
                })

            respuesta = {
                "status": "OK", "fuente": "BD_PRINCIPAL",
                "total": len(resultados), "resultados": resultados
            }

        elif tipo == "CONSULTA_INTERSECCION":
            # Busco los últimos datos de una intersección específica
            inter = consulta.get("interseccion", "")

            cursor.execute("""
                SELECT tipo_sensor, estado_trafico, Q, Vp, D, timestamp_procesado
                FROM eventos_trafico
                WHERE interseccion = ?
                ORDER BY timestamp_procesado DESC
                LIMIT 10
            """, (inter,))

            resultados = []
            for fila in cursor.fetchall():
                resultados.append({
                    "tipo_sensor": fila[0], "estado_trafico": fila[1],
                    "Q": fila[2], "Vp": fila[3], "D": fila[4],
                    "timestamp": fila[5]
                })

            respuesta = {
                "status": "OK", "fuente": "BD_PRINCIPAL",
                "interseccion": inter,
                "total": len(resultados), "resultados": resultados
            }

        elif tipo == "CONSULTA_ESTADOS":
            # Resumen: cuántos eventos hay por cada estado de tráfico
            cursor.execute("""
                SELECT estado_trafico, COUNT(*) FROM eventos_trafico
                GROUP BY estado_trafico ORDER BY COUNT(*) DESC
            """)
            resumen = {}
            for fila in cursor.fetchall():
                resumen[fila[0]] = fila[1]

            respuesta = {
                "status": "OK", "fuente": "BD_PRINCIPAL",
                "resumen_estados": resumen
            }

        elif tipo == "CONSULTA_THROUGHPUT":
            # Total de registros en la BD (para medir rendimiento)
            total = cursor.execute("SELECT COUNT(*) FROM eventos_trafico").fetchone()[0]
            respuesta = {
                "status": "OK", "fuente": "BD_PRINCIPAL",
                "total_registros": total
            }

        else:
            respuesta = {"status": "ERROR", "mensaje": f"No entiendo: {tipo}"}

        conn.close()
        socket.send_json(respuesta)
        print(f"[BD-PRINCIPAL] Respuesta enviada")


def main():
    print("=" * 60)
    print("  BASE DE DATOS PRINCIPAL - PC3 (Persistencia)")
    print("=" * 60)
    print(f"  PULL datos:    tcp://{BD_IP}:{PUERTO_PULL}")
    print(f"  REP consultas: tcp://{BD_IP}:{PUERTO_REP}")
    print(f"  Archivo BD:    {BD_ARCHIVO}")
    print("=" * 60)

    # Creo las tablas si no existen
    crear_tablas()

    contexto = zmq.Context()

    # Hilo para recibir datos de la analítica
    t1 = threading.Thread(target=hilo_recibir_datos, args=(contexto,), daemon=True)
    t1.start()

    # Hilo para atender consultas del monitoreo
    t2 = threading.Thread(target=hilo_consultas, args=(contexto,), daemon=True)
    t2.start()

    print("[BD-PRINCIPAL] Servicio corriendo. Ctrl+C para detener.\n")

    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        print("\n[BD-PRINCIPAL] Cerrando...")
        print("[BD-PRINCIPAL] Listo.")


if __name__ == "__main__":
    main()

File: java/pc3/src/main/java/BdPrincipal.java
// bdprincipal.java - base de datos principal - pc3
//
// este servicio es la bd principal del sistema
// recibe datos procesados de la analitica pull y los guarda en sqlite
// ademas expone un socket rep para consultas del monitoreo y sincronizacion
//
// autores: miguel angel acuna, juan david acuna, y samuel felipe manrique - sistemas distribuidos 2026-10

import org.zeromq.SocketType;
import org.zeromq.ZContext;
import org.zeromq.ZMQ;
import org.json.JSONObject;
import org.json.JSONArray;

import java.sql.*;
import java.time.Instant;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;

public class BdPrincipal {

    // configuracion de red
    // ============================================================
    static String BD_IP = "10.43.99.183";   // pc3
    static int PUERTO_PULL = 5570;          // puerto para recibir datos de analitica
    static int PUERTO_REP = 5571;           // puerto para responder consultas y sincronizacion

    static String BD_ARCHIVO = "trafico.db";   // nombre del archivo sqlite

    static String timestampAhora() {
        return DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss'Z'")
                .withZone(ZoneOffset.UTC)
                .format(Instant.now());
    }

    // creacion de tablas si no existen
    static void crearTablas() {
        try (Connection conn = DriverManager.getConnection("jdbc:sqlite:" + BD_ARCHIVO)) {
            Statement stmt = conn.createStatement();

            // tabla principal de eventos de trafico
            stmt.execute("CREATE TABLE IF NOT EXISTS eventos_trafico ("
                    + "id INTEGER PRIMARY KEY AUTOINCREMENT,"
                    + "interseccion TEXT, tipo_sensor TEXT, datos_sensor TEXT,"
                    + "estado_trafico TEXT, Q REAL, Vp REAL, D REAL, via TEXT,"
                    + "timestamp_procesado TEXT, timestamp_insercion TEXT)");

            // tabla para guardar acciones de control semaforos/emergencia
            stmt.execute("CREATE TABLE IF NOT EXISTS acciones_control ("
                    + "id INTEGER PRIMARY KEY AUTOINCREMENT,"
                    + "interseccion TEXT, via TEXT, tipo_accion TEXT, detalles TEXT, timestamp TEXT)");

            System.out.println("[BD-PRINCIPAL] Base de datos '" + BD_ARCHIVO + "' lista");
        } catch (SQLException e) {
            System.out.println("[BD-PRINCIPAL] Error creando tablas: " + e.getMessage());
        }
    }

    // guarda un evento procesado en la bd
    static int guardarEvento(JSONObject registro) {
        int total = 0;
        try (Connection conn = DriverManager.getConnection("jdbc:sqlite:" + BD_ARCHIVO)) {
            PreparedStatement ps = conn.prepareStatement(
                    "INSERT INTO eventos_trafico "
                    + "(interseccion, tipo_sensor, datos_sensor, estado_trafico, Q, Vp, D, via, "
                    + "timestamp_procesado, timestamp_insercion) VALUES (?,?,?,?,?,?,?,?,?,?)");

            ps.setString(1, registro.optString("interseccion", ""));
            ps.setString(2, registro.optString("tipo_sensor", ""));
            ps.setString(3, registro.optJSONObject("datos_sensor") != null
                    ? registro.getJSONObject("datos_sensor").toString() : "{}");
            ps.setString(4, registro.optString("estado_trafico", ""));
            ps.setDouble(5, registro.optDouble("Q", 0));
            ps.setDouble(6, registro.optDouble("Vp", 0));
            ps.setDouble(7, registro.optDouble("D", 0));
            ps.setString(8, registro.optString("via", ""));
            ps.setString(9, registro.optString("timestamp_procesado", ""));
            ps.setString(10, timestampAhora());
            ps.executeUpdate();

            ResultSet rs = conn.createStatement().executeQuery("SELECT COUNT(*) FROM eventos_trafico");
            if (rs.next()) total = rs.getInt(1);
        } catch (SQLException e) {
            System.out.println("[BD-PRINCIPAL] Error guardando: " + e.getMessage());
        }
        return total;
    }

    // guarda una accion de control emergencia/ola verde en la bd
    static void guardarAccionControl(JSONObject accion) {
        try (Connection conn = DriverManager.getConnection("jdbc:sqlite:" + BD_ARCHIVO)) {
            PreparedStatement ps = conn.prepareStatement(
                    "INSERT INTO acciones_control (interseccion, via, tipo_accion, detalles, timestamp) VALUES (?,?,?,?,?)");
            ps.setString(1, accion.optString("interseccion", ""));
            ps.setString(2, accion.optString("via", ""));
            ps.setString(3, accion.optString("tipo_accion", ""));
            ps.setString(4, accion.optString("detalles", ""));
            ps.setString(5, accion.optString("timestamp", timestampAhora()));
            ps.executeUpdate();
            System.out.println("[BD-PRINCIPAL] Accion de control guardada");
        } catch (SQLException e) {
            System.out.println("[BD-PRINCIPAL] Error guardando accion de control: " + e.getMessage());
        }
    }

    // hilo que recibe datos procesados de la analitica pull
    static void hiloRecibirDatos(ZContext contexto) {
        ZMQ.Socket socket = contexto.createSocket(SocketType.PULL);
        socket.bind("tcp://" + BD_IP + ":" + PUERTO_PULL);

        ZMQ.Poller poller = contexto.createPoller(1);
        poller.register(socket, ZMQ.Poller.POLLIN);

        System.out.println("[BD-PRINCIPAL] PULL esperando datos en tcp://" + BD_IP + ":" + PUERTO_PULL);

        while (!Thread.currentThread().isInterrupted()) {
            poller.poll(2000);
            if (!poller.pollin(0)) continue;

            String msg = socket.recvStr();
            if (msg == null) continue;

            try {
                JSONObject registro = new JSONObject(msg);
                int total = guardarEvento(registro);
                String inter = registro.optString("interseccion", "?");
                String via = registro.optString("via", "?");
                String estado = registro.optString("estado_trafico", "?");
                System.out.println("[BD-PRINCIPAL] Guardado PULL: " + inter + " (" + via + ") | " + estado + " | Total: " + total);
            } catch (Exception e) {
                System.out.println("[BD-PRINCIPAL] Error procesando mensaje PULL: " + e.getMessage());
            }
        }
    }

    // responde a las solicitudes de consulta monitoreo y sincronizacion
    static void hiloResponderConsultas(ZContext contexto) {
        ZMQ.Socket socket = contexto.createSocket(SocketType.REP);
        socket.bind("tcp://" + BD_IP + ":" + PUERTO_REP);

        System.out.println("[BD-PRINCIPAL] REP esperando consultas en tcp://" + BD_IP + ":" + PUERTO_REP);

        while (!Thread.currentThread().isInterrupted()) {
            String msg = socket.recvStr();
            if (msg == null) continue;

            JSONObject respuesta = new JSONObject();
            try {
                JSONObject peticion = new JSONObject(msg);
                String accion = peticion.optString("accion", "");

                System.out.println("[BD-PRINCIPAL] Peticion REP recibida: " + accion);

                if (accion.equals("ULTIMO_REGISTRO")) {
                    // obtiene el ultimo timestamp y el id maximo de la bd
                    try (Connection conn = DriverManager.getConnection("jdbc:sqlite:" + BD_ARCHIVO)) {
                        Statement stmt = conn.createStatement();
                        ResultSet rs = stmt.executeQuery("SELECT MAX(id), MAX(timestamp_procesado) FROM eventos_trafico");
                        if (rs.next()) {
                            respuesta.put("ultimo_id", rs.getInt(1));
                            respuesta.put("ultimo_timestamp", rs.getString(2) != null ? rs.getString(2) : "1970-01-01T00:00:00Z");
                        } else {
                            respuesta.put("ultimo_id", 0);
                            respuesta.put("ultimo_timestamp", "1970-01-01T00:00:00Z");
                        }
                    }
                    respuesta.put("resultado", "OK");

                } else if (accion.equals("INSERTAR_REGISTRO")) {
                    JSONObject registro = peticion.getJSONObject("registro");
                    guardarEvento(registro);
                    respuesta.put("resultado", "OK");
                    System.out.println("[BD-PRINCIPAL] Registro de sincronización insertado");

                } else if (accion.equals("GUARDAR_ACCION_CONTROL")) {
                    JSONObject control = peticion.getJSONObject("control");
                    guardarAccionControl(control);
                    respuesta.put("resultado", "OK");

                } else if (accion.equals("CONSULTA_HISTORICA")) {
                    String inicio = peticion.optString("inicio", "");
                    String fin = peticion.optString("fin", "");
                    JSONArray eventos = new JSONArray();

                    try (Connection conn = DriverManager.getConnection("jdbc:sqlite:" + BD_ARCHIVO)) {
                        PreparedStatement ps = conn.prepareStatement(
                                "SELECT interseccion, via, tipo_sensor, estado_trafico, Q, Vp, D, timestamp_procesado "
                                + "FROM eventos_trafico WHERE timestamp_procesado BETWEEN ? ?");
                        // nota: corregir sintaxis sql usaremos between ? and ?
                        PreparedStatement psFixed = conn.prepareStatement(
                                "SELECT interseccion, via, tipo_sensor, estado_trafico, Q, Vp, D, timestamp_procesado "
                                + "FROM eventos_trafico WHERE timestamp_procesado BETWEEN ? AND ?");
                        psFixed.setString(1, inicio);
                        psFixed.setString(2, fin);
                        ResultSet rs = psFixed.executeQuery();

                        while (rs.next()) {
                            JSONObject ev = new JSONObject();
                            ev.put("interseccion", rs.getString(1));
                            ev.put("via", rs.getString(2));
                            ev.put("tipo_sensor", rs.getString(3));
                            ev.put("estado_trafico", rs.getString(4));
                            ev.put("Q", rs.getDouble(5));
                            ev.put("Vp", rs.getDouble(6));
                            ev.put("D", rs.getDouble(7));
                            ev.put("timestamp_procesado", rs.getString(8));
                            eventos.put(ev);
                        }
                    }
                    respuesta.put("resultado", "OK");
                    respuesta.put("eventos", eventos);
                    System.out.println("[BD-PRINCIPAL] Consulta histórica devuelta: " + eventos.length() + " registros");

                } else if (accion.equals("CONSULTA_PUNTUAL")) {
                    String interseccion = peticion.optString("interseccion", "");
                    JSONArray eventos = new JSONArray();

                    try (Connection conn = DriverManager.getConnection("jdbc:sqlite:" + BD_ARCHIVO)) {
                        PreparedStatement ps = conn.prepareStatement(
                                "SELECT interseccion, via, tipo_sensor, estado_trafico, Q, Vp, D, timestamp_procesado "
                                + "FROM eventos_trafico WHERE interseccion = ? ORDER BY id DESC LIMIT 5");
                        ps.setString(1, interseccion);
                        ResultSet rs = ps.executeQuery();

                        while (rs.next()) {
                            JSONObject ev = new JSONObject();
                            ev.put("interseccion", rs.getString(1));
                            ev.put("via", rs.getString(2));
                            ev.put("tipo_sensor", rs.getString(3));
                            ev.put("estado_trafico", rs.getString(4));
                            ev.put("Q", rs.getDouble(5));
                            ev.put("Vp", rs.getDouble(6));
                            ev.put("D", rs.getDouble(7));
                            ev.put("timestamp_procesado", rs.getString(8));
                            eventos.put(ev);
                        }
                    }
                    respuesta.put("resultado", "OK");
                    respuesta.put("eventos", eventos);
                    System.out.println("[BD-PRINCIPAL] Consulta puntual devuelta: " + eventos.length() + " registros");

                } else {
                    respuesta.put("resultado", "ERROR");
                    respuesta.put("mensaje", "Accion no soportada");
                }
            } catch (Exception e) {
                respuesta.put("resultado", "ERROR");
                respuesta.put("mensaje", e.getMessage());
            }

            socket.send(respuesta.toString());
        }
    }

    public static void main(String[] args) {
        System.out.println("============================================================");
        System.out.println("  BASE DE DATOS PRINCIPAL - PC3 (Persistencia)");
        System.out.println("============================================================");
        System.out.println("  PULL datos:    tcp://" + BD_IP + ":" + PUERTO_PULL);
        System.out.println("  REP consultas: tcp://" + BD_IP + ":" + PUERTO_REP);
        System.out.println("  Archivo BD:    " + BD_ARCHIVO);
        System.out.println("============================================================");

        crearTablas();

        ZContext contexto = new ZContext();

        // hilo pull para recibir de analitica
        Thread t1 = new Thread(() -> hiloRecibirDatos(contexto));
        t1.setDaemon(true);
        t1.start();

        // hilo rep para consultas de monitoreo
        Thread t2 = new Thread(() -> hiloResponderConsultas(contexto));
        t2.setDaemon(true);
        t2.start();

        System.out.println("[BD-PRINCIPAL] Servicio corriendo. Ctrl+C para detener.\n");

        try {
            while (true) { Thread.sleep(1000); }
        } catch (InterruptedException e) {
            System.out.println("\n[BD-PRINCIPAL] Cerrando...");
            contexto.close();
            System.out.println("[BD-PRINCIPAL] Listo.");
        }
    }
}
